ns for fx: make weights comparison work on N models

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #52402 ns for fx: update graph matching to not match nodes with equal types
* #52395 ns for fx: support linear_relu for weight matching
* #52368 ns for fx: allow graph matching of parents of cat
* #52357 ns for fx: make unshadowed activation comparison work for N models
* **#52356 ns for fx: make weights comparison work on N models**
* #52130 ns for fx: add support for subgraph matching
* #52092 NS for FX: add test for a simple sparsenn model

Summary:

Refactor the NS for FX compare weights API to be able to
work on N models and do arbitrary matching strategies.

We factor out a util which takes a model and a list of
nodes to extract weights for, with names to give the extracted
weights.  The user can then call this util with a set
of nodes and names created in any way they want.

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26487271](https://our.internmc.facebook.com/intern/diff/D26487271)

vkuzo committed:
ns for fx: make weights comparison work on N models

Summary:

Refactor the NS for FX compare weights API to be able to
work on N models and do arbitrary matching strategies.

We factor out a util which takes a model and a list of
nodes to extract weights for, with names to give the extracted
weights.  The user can then call this util with a set
of nodes and names created in any way they want.

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
rebase on "ns for fx: make weights comparison work on N models"


Summary:

Refactor the NS for FX compare weights API to be able to
work on N models and do arbitrary matching strategies.

We factor out a util which takes a model and a list of
nodes to extract weights for, with names to give the extracted
weights.  The user can then call this util with a set
of nodes and names created in any way they want.

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26487271](https://our.internmc.facebook.com/intern/diff/D26487271)

[ghstack-poisoned]

raghuramank100 commented on torch/quantization/ns/numeric_suite_core_apis_fx.py:
```diff
@@ -27,65 +27,71 @@
+    model_name_a: str,
+    gm_a: GraphModule,
+    model_name_b: str,
+    gm_b: GraphModule,
```
For a future PR: We need to provide flexibility to users in terms of specifying modules that they want to compare both in terms of module names and in terms of module types.

vkuzo replied:
agreed, saving that for a future PR

vkuzo committed:
rebase on "ns for fx: make weights comparison work on N models"


Summary:

Refactor the NS for FX compare weights API to be able to
work on N models and do arbitrary matching strategies.

We factor out a util which takes a model and a list of
nodes to extract weights for, with names to give the extracted
weights.  The user can then call this util with a set
of nodes and names created in any way they want.

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26487271](https://our.internmc.facebook.com/intern/diff/D26487271)

[ghstack-poisoned]

vkuzo committed:
fix asan on "ns for fx: make weights comparison work on N models"


Summary:

Refactor the NS for FX compare weights API to be able to
work on N models and do arbitrary matching strategies.

We factor out a util which takes a model and a list of
nodes to extract weights for, with names to give the extracted
weights.  The user can then call this util with a set
of nodes and names created in any way they want.

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26487271](https://our.internmc.facebook.com/intern/diff/D26487271)

[ghstack-poisoned]
